[coor-slicing] DeviceMesh.is_current_rank_part_of_mesh

aorenste opened the pull request:
Adds two methods to DeviceMesh:
- `is_current_rank_part_of_mesh`
    There are a number of places where we only care if our current rank is part of the DeviceMesh but don't actually care about other information. So instead of getting all the mesh coordinates and checking for `None` we can just have a predicate that says whether we are part of the rank or not.
- `sym_get_coordinate`
    Morally equivalent to `get_coordinate()[i]` - Instead of getting all the mesh coordinates as a list and extracting the one we want this allows specifying the rank and just getting the specific coordinate for that rank. Right now it only returns `int` but in the future can also return a `SymInt`.

Today both of these are a simple lookup in the `_coordinate_on_dim` array but a later PR will specialize them to properly limit their scope to make compile-on-one-rank happier.

Also the LocalTensorMode ad-hoc method patching was becoming unwieldy (and I had to add more items to it) so I made it automated based off a list instead of easy to miss one-offs.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169552
* #169551
* #169550
* #169549
* __->__ #169548



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

aorenste committed:
WIP: DeviceMesh.is_part_of_mesh

[ghstack-poisoned]

aorenste committed:
Update on "WIP: DeviceMesh.is_part_of_mesh"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: DeviceMesh.is_part_of_mesh"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: DeviceMesh.is_part_of_mesh"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: DeviceMesh.is_part_of_mesh"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: DeviceMesh.is_part_of_mesh"




[ghstack-poisoned]

aorenste committed:
Update on "DeviceMesh.is_current_rank_part_of_mesh"

Adds two methods to DeviceMesh:
- `is_current_rank_part_of_mesh`
    There are a number of places where we only care if our current rank is part of the DeviceMesh but don't actually care about other information. So instead of getting all the mesh coordinates and checking for `None` we can just have a predicate that says whether we are part of the rank or not.
- `sym_get_coordinate`
    Morally equivalent to `get_coordinate()[i]` - Instead of getting all the mesh coordinates as a list and extracting the one we want this allows specifying the rank and just getting the specific coordinate for that rank. Right now it only returns `int` but in the future can also return a `SymInt`.

Today both of these are a simple lookup in the `_coordinate_on_dim` array but a later PR will specialize them to properly limit their scope to make compile-on-one-rank happier.




[ghstack-poisoned]

aorenste committed:
Update on "DeviceMesh.is_current_rank_part_of_mesh"

Adds two methods to DeviceMesh:
- `is_current_rank_part_of_mesh`
    There are a number of places where we only care if our current rank is part of the DeviceMesh but don't actually care about other information. So instead of getting all the mesh coordinates and checking for `None` we can just have a predicate that says whether we are part of the rank or not.
- `sym_get_coordinate`
    Morally equivalent to `get_coordinate()[i]` - Instead of getting all the mesh coordinates as a list and extracting the one we want this allows specifying the rank and just getting the specific coordinate for that rank. Right now it only returns `int` but in the future can also return a `SymInt`.

Today both of these are a simple lookup in the `_coordinate_on_dim` array but a later PR will specialize them to properly limit their scope to make compile-on-one-rank happier.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "DeviceMesh.is_current_rank_part_of_mesh"

Adds two methods to DeviceMesh:
- `is_current_rank_part_of_mesh`
    There are a number of places where we only care if our current rank is part of the DeviceMesh but don't actually care about other information. So instead of getting all the mesh coordinates and checking for `None` we can just have a predicate that says whether we are part of the rank or not.
- `sym_get_coordinate`
    Morally equivalent to `get_coordinate()[i]` - Instead of getting all the mesh coordinates as a list and extracting the one we want this allows specifying the rank and just getting the specific coordinate for that rank. Right now it only returns `int` but in the future can also return a `SymInt`.

Today both of these are a simple lookup in the `_coordinate_on_dim` array but a later PR will specialize them to properly limit their scope to make compile-on-one-rank happier.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zpcore reviewed (commented):
Thanks, I hope I can approve this :) Let's check with @fduwjj on his feedbacks once he is back.

zpcore reviewed (approved):
LGTM!

aorenste committed:
Update on "[coor-slicing] DeviceMesh.is_current_rank_part_of_mesh"

Adds two methods to DeviceMesh:
- `is_current_rank_part_of_mesh`
    There are a number of places where we only care if our current rank is part of the DeviceMesh but don't actually care about other information. So instead of getting all the mesh coordinates and checking for `None` we can just have a predicate that says whether we are part of the rank or not.
- `sym_get_coordinate`
    Morally equivalent to `get_coordinate()[i]` - Instead of getting all the mesh coordinates as a list and extracting the one we want this allows specifying the rank and just getting the specific coordinate for that rank. Right now it only returns `int` but in the future can also return a `SymInt`.

Today both of these are a simple lookup in the `_coordinate_on_dim` array but a later PR will specialize them to properly limit their scope to make compile-on-one-rank happier.

Also the LocalTensorMode ad-hoc method patching was becoming unwieldy (and I had to add more items to it) so I made it automated based off a list instead of easy to miss one-offs.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] DeviceMesh.is_current_rank_part_of_mesh"

Adds two methods to DeviceMesh:
- `is_current_rank_part_of_mesh`
    There are a number of places where we only care if our current rank is part of the DeviceMesh but don't actually care about other information. So instead of getting all the mesh coordinates and checking for `None` we can just have a predicate that says whether we are part of the rank or not.
- `sym_get_coordinate`
    Morally equivalent to `get_coordinate()[i]` - Instead of getting all the mesh coordinates as a list and extracting the one we want this allows specifying the rank and just getting the specific coordinate for that rank. Right now it only returns `int` but in the future can also return a `SymInt`.

Today both of these are a simple lookup in the `_coordinate_on_dim` array but a later PR will specialize them to properly limit their scope to make compile-on-one-rank happier.

Also the LocalTensorMode ad-hoc method patching was becoming unwieldy (and I had to add more items to it) so I made it automated based off a list instead of easy to miss one-offs.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
